site/en/docs/extensions/mv3/messaging/index.md

```diff
@@ -289,59 +283,65 @@ Extensions [can exchange messages][native-client] with native applications that
-sanitize all input][security-santize]. Assume any data sent to the content script might leak to the web page.
+[Content scripts are less trustworthy][security-cs] than the extension service worker.
+For example, a malicious webpage might be able to compromise the rendering process that
+runs the content scripts. Assume that messages from a content script might have been
+crafted by an attacker and make sure to [validate and sanitize all input][security-santize].
+Assume any data sent to the content script might leak to the web page.
 Limit the scope of privileged actions that can be triggered by messages received from content
 scripts.
 
 ### Cross-site scripting {: #cross-site-scripting }
 
-When receiving a message from a content script or another extension, your scripts should be careful
-not to fall victim to [cross-site scripting][wiki-cross-site]. This advice applies to scripts running inside the
-extension background page as well as to content scripts running inside other web origins.
-Specifically, avoid using dangerous APIs such as the ones below:
+When receiving a message from a content script or another extension, make sure to protect
+your scripts against [cross-site scripting][wiki-cross-site]. This advice applies to scripts running inside the
+extension background page as well as to content scripts running in other web origins.
+
+{% Compare 'better', 'Safer methods' %}
+Use APIs that don't run scripts whenever possible:
 
-{% Label %}content-script.js or service-worker.js{% endLabel %}
+{% Label %}service-worker.js{% endLabel %}
 
 ```js
 chrome.tabs.sendMessage(tab.id, {greeting: "hello"}, function(response) {
-  // WARNING! Might be evaluating an evil script!
-  var resp = eval("(" + response.farewell + ")");
+  // JSON.parse doesn't evaluate the attacker's scripts.
+  var resp = JSON.parse(response.farewell);
 });
 ```
 
-{% Label %}content-script.js or service-worker.js{% endLabel %}
+{% Label %}service-worker.js{% endLabel %}
 
 ```js
 chrome.tabs.sendMessage(tab.id, {greeting: "hello"}, function(response) {
-  // WARNING! Might be injecting a malicious script!
-  document.getElementById("resp").innerHTML = response.farewell;
+  // innerText does not let the attacker inject HTML elements.
+  document.getElementById("resp").innerText = response.farewell;
 });
 ```
+{% endCompare %}
 
-Instead, prefer safer APIs that do not run scripts:
+{% Compare 'worse', 'Unsafe methods' %}
+Avoid using the following methods that make your extension vulnerable:
 
-{% Label %}content-script.js or service-worker.js{% endLabel %}
+{% Label %}service-worker.js{% endLabel %}
 
 ```js
 chrome.tabs.sendMessage(tab.id, {greeting: "hello"}, function(response) {
-  // JSON.parse does not evaluate the attacker's scripts.
-  var resp = JSON.parse(response.farewell);
+  // WARNING! Might be evaluating a malicious script!
+  var resp = eval("(" + response.farewell + ")");
```

Comment:
@oliverdunk Should we update the "unsafe" code samples, since MV3 doesn't allow eval() or access to the dom?

Reply:
I considered this, but I'm not sure of a better example and even if it isn't allowed people still write it, so I'm not too concerned.

Reply:
This is specifically an example of a bad way to do something :)

Reply:
Are their people in our audience who will see the examples presented, and go "okay, what about Joe's example? Is that safe?" If that's a thing we can expect our readers to think, it might be worth mentioning.

Reply:
I think the concatenation is actually causing the confusion here - it's not an important part of the snippet. Maybe something like this would be clearer...

Bad:

```
const script = 'alert(' + response.farewell + ')';
eval(script);
```

Good:

```
alert(response.farewell)
```

Reply:
OK. I forgot to look at the context when I asked the question about `eval()`. It makes perfect sense as the antipattern. @oliverdunk, I think your proposal is right, but we need to use modern programming idioms.